Mirror Services (#5)
COSI-Lab/book: src/mirror/services.md

# Services

There are some services on Mirror whose documentation does not fit well in the documentation of the software project. These are mostly configuration and documentation about services running on Mirror. 

## NGINX

_updated: Jan 15th 2023_ 

> [NGINX](https://nginx.org/en/) is Mirror's web server listening on ports 80 (http) and 443 (https). NGINX is responisbile for servering project files, generating index pages (for example <https://mirror.clarkson.edu/blender>), and redirecting requests for the website to [Mirror Software](./software.md).

Reviewer: ```suggestion
[NGINX](https://nginx.org/en/) is Mirror's web server listening on ports 80 (http) and 443 (https). NGINX is responsible for serving project files, generating index pages (for example <https://mirror.clarkson.edu/blender>), and redirecting requests for the website to [Mirror Software](./software.md).
```

	location ~ ^/(home|history|stats|projects|sync|favicon.ico|map|health|css/|img/|js/|fonts/|api/) {
		proxy_pass http://127.0.0.1:8012$request_uri;
	}

	# Relocation of linuxmint things
	rewrite ^/linuxmint/iso/images/(.*)$ /linuxmint-images/$1 permanent;
	rewrite ^/linuxmint/packages/(.*)$ /linuxmint-packages/$1 permanent;

	# Allow localhost scrapping of status
	location = /status {
		allow 127.0.0.1;
		stub_status;
	}

	# There is this very strange behavior where IP's located in China are very interested in these two files 
	# This useless traffic accounts for around 15% of all requests to Mirror. Nothing we have tried has effectively stopped this traffic.
	# Returning 444 tell NGINX to ignore these connections as soon as possible (there is no point in sending a 404 page)
	location = /centos/8.5.2111/isos/aarch64/CentOS-8.5.2111-aarch64-dvd1.iso {
		return 444;
	}

	location = /centos/8.5.2111/isos/x86_64/CentOS-8.5.2111-x86_64-dvd1.iso {
		return 444;
	}

	# Track access to these odd locations in another file. Maybe we could IP-Ban these connections.
	access_log /var/log/nginx/ban.log combined if=$ban;
}
```

## Influxdb

_updated: Jan 15th 2023_ 

> [Influxdb](https://www.influxdata.com/products/influxdb-overview/) is an open source [time-series database](https://en.wikipedia.org/wiki/Time_series_database). [Mirror Software](./software.md) will process our [NGINX](#nginx) log files extracting interesting information, aggrate it, and then store the results in Influxdb. 

Reviewer: ```suggestion
[Influxdb](https://www.influxdata.com/products/influxdb-overview/) is an open source [time-series database](https://en.wikipedia.org/wiki/Time_series_database). [Mirror Software](./software.md) will process our [NGINX](#nginx) log files extracting interesting information, aggregating it, and then storing the results in Influxdb. 
```

  token = "REDACTED"

  ## Organization is the name of the organization you wish to write to; must exist.
  organization = "COSI"

  ## Destination bucket to write into.
  bucket = "system"

  ## Optional TLS Config for use on HTTP connections.
  # tls_ca = "/etc/telegraf/ca.pem"
  # tls_cert = "/etc/telegraf/cert.pem"
  # tls_key = "/etc/telegraf/key.pem"

  ## Use TLS but skip chain & host verification
  insecure_skip_verify = true
[[inputs.cpu]]
  ## Whether to report per-cpu stats or not
  percpu = true

  ## Whether to report total system cpu stats or not
  totalcpu = true

  ## If true, collect raw CPU time metrics
  collect_cpu_time = false

  ## If true, compute and report the sum of all non-idle CPU states
  report_active = false
[[inputs.disk]]
  ## Ignore mount points by filesystem type.
  ignore_fs = ["tmpfs", "devtmpfs", "devfs", "iso9660", "overlay", "aufs", "squashfs"]
[[inputs.diskio]]
[[inputs.mem]]
[[inputs.net]]
[[inputs.nginx]]
  # An array of Nginx stub_status URI to gather stats.
  urls = ["http://localhost/status"]

  # HTTP response timeout (default: 5s)
  response_timeout = "5s"
[[inputs.system]]
[[inputs.zfs]]
```

## RKHunter

_updated: Nov 6th 2022_

> [RKHunter](https://rkhunter.sourceforge.net/) or "Rootkit Hunter" is a script that checks linux systems for signs of known linux rootkits. Assuming other security measures are well followed we should not expect Mirror to become infected with a rootkit. It might be worth an investment in running the tool automaticaaly (maybe daily?) but we don't currently have satisfactory methods of sending mail when warnings occur. Until then (and maybe forever) here is some instructions on manually running RKHunter. 

Reviewer: ```suggestion
[RKHunter](https://rkhunter.sourceforge.net/) or "Rootkit Hunter" is a script that checks linux systems for signs of known linux rootkits. Assuming other security measures are well followed we don't expect Mirror to become infected with a rootkit. It might be worth running the tool automatically (maybe daily?), however, we don't currently have a satisfactory method of sending mail when warnings occur. Until then (and maybe forever) here are some instructions for manually running RKHunter. 
```

**Install:**

```
sudo apt install rkhunter
```

**Update:**

This command tells RKHunter to check for newly added software packages. We don't install software often, but it isn't a bad idea to run this event once in a while.

```
sudo rkhunter --propupd
```

**Usage:**

```
sudo rkhunter --check --skip-keypress --report-warnings-only
```

| Flag | Meaning |
| :--: | :--- |
| `--checkall` | Runs all tests. |
| `--skip-keypress` | RKHunter has a pretty annoying "press ENTER for the next scan". This disables that. |
| `--report-warnings-only` | Recommended when using `--skip-keypress` otherwise it's easy to miss results. |

**False positives:**

```
Warning: The command '/usr/bin/lwp-request' has been replaced by a script: /usr/bin/lwp-request: Perl script text executable
```

> This seems to be a wellknown issue on debian based machines. 2 attempts have been made to stop this error without success.

Reviewer: ```suggestion
This seems to be a well known issue on Debian based machines. So far two attempts have been made to stop this error, both without success.
```
